metapool/sample_sheet.py

```diff
@@ -0,0 +1,596 @@
+import re
+import csv
+import collections
+import warnings
+
+from datetime import datetime
+from itertools import chain, repeat, islice
+
+import sample_sheet
+import pandas as pd
+
+from metapool.metapool import (bcl_scrub_name, sequencer_i5_index,
+                               REVCOMP_SEQUENCERS)
+from metapool.plate import ErrorMessage, WarningMessage
+
+_KL_SAMPLE_SHEET_SECTIONS = [
+    'Header', 'Reads', 'Settings', 'Data', 'Bioinformatics', 'Contact'
+]
+
+_KL_SAMPLE_SHEET_DATA_COLUMNS = [
+    'Sample_ID', 'Sample_Name', 'Sample_Plate', 'Sample_Well', 'I7_Index_ID',
+    'index', 'I5_Index_ID', 'index2', 'Sample_Project', 'Well_description'
+]
+
+_KL_AMPLICON_REMAPPER = {
+    'sample sheet Sample_ID': 'Sample_ID',
+    'Sample': 'Sample_Name',
+    'Project Plate': 'Sample_Plate',
+    'Well': 'Sample_Well',
+    'Name': 'I7_Index_ID',
+    'Golay Barcode': 'index',
+    'Project Name': 'Sample_Project',
+}
+
+_KL_METAGENOMICS_REMAPPER = {
+    'sample sheet Sample_ID': 'Sample_ID',
+    'Sample': 'Sample_Name',
+    'Project Plate': 'Sample_Plate',
+    'Well': 'Sample_Well',
+    'i7 name': 'I7_Index_ID',
+    'i7 sequence': 'index',
+    'i5 name': 'I5_Index_ID',
+    'i5 sequence': 'index2',
+    'Project Name': 'Sample_Project',
+}
+
+_AMPLICON = 'TruSeq HT'
+_METAGENOMICS = 'Metagenomics'
+_ASSAYS = {_AMPLICON, _METAGENOMICS}
+
+_READS = {
+    'Read1': 151,
+    'Read2': 151
+}
+
+_SETTINGS = {
+    'ReverseComplement': 0
+}
+
+_HEADER = {
+    'IEMFileVersion': 4,
+    'Investigator Name': 'Knight',
+    'Experiment Name': 'RKL_experiment',
+    'Date': None,
+    'Workflow': 'GenerateFASTQ',
+    'Application': 'FASTQ Only',
+    'Assay': None,
+    'Description': '',
+    'Chemistry': 'Default',
+}
+
+_BIOINFORMATICS_COLUMNS = {
+    'Sample_Project', 'QiitaID', 'BarcodesAreRC', 'ForwardAdapter',
+    'ReverseAdapter', 'HumanFiltering'
+}
+
+_CONTACT_COLUMNS = {
+    'Sample_Project', 'Email'
+}
+
+_BIOINFORMATICS_AND_CONTACT = {
+    'Bioinformatics': None,
+    'Contact': None
+}
+
+_ALL_METADATA = {**_HEADER, **_SETTINGS, **_READS,
+                 **_BIOINFORMATICS_AND_CONTACT}
+
+
+class KLSampleSheet(sample_sheet.SampleSheet):
+    def __init__(self, path=None):
+        """Knight Lab's SampleSheet subclass
+
+        Includes a number of parsing and writing changes to allow for the
+        inclusion of the Bioinformatics and Contact sections.
+
+        The majority of the code in the _parse and write methods are copied
+        from release 0.11.0 https://github.com/clintval/sample-sheet/blob/\
+c3df258541a384a5058f8aa46b343ff032d8e247/sample_sheet/__init__.py
```

Comment:
That 404 may have been because of the backslash. But in any case I've created a tinyurl that I'll use to replace for this long URL. I wanted to make sure I included the git SHA, otherwise the blob/master URL will lead to the most recent version which might have not been the one I used to base this code on.

https://tinyurl.com/clintval-sample-sheet